packages/playwright-core/src/utils/comparators.ts

```diff
@@ -56,13 +56,16 @@ function compareImages(mimeType: string, actualBuffer: Buffer | string, expected
 
   let actual: ImageData = mimeType === 'image/png' ? PNG.sync.read(actualBuffer) : jpegjs.decode(actualBuffer, { maxMemoryUsageInMB: JPEG_JS_MAX_BUFFER_SIZE_IN_MB });
   let expected: ImageData = mimeType === 'image/png' ? PNG.sync.read(expectedBuffer) : jpegjs.decode(expectedBuffer, { maxMemoryUsageInMB: JPEG_JS_MAX_BUFFER_SIZE_IN_MB });
-  const size = { width: Math.max(expected.width, actual.width), height: Math.max(expected.height, actual.height) };
+  const size = { width: Math.min(expected.width, actual.width), height: Math.min(expected.height, actual.height) };
   let sizesMismatchError = '';
   if (expected.width !== actual.width || expected.height !== actual.height) {
-    sizesMismatchError = `Expected an image ${expected.width}px by ${expected.height}px, received ${actual.width}px by ${actual.height}px. `;
+    const maxDiffSize = 1 + (options.maxDiffSize ?? 1);
+    if (Math.abs(expected.width - actual.width) >= maxDiffSize || Math.abs(expected.height - actual.height) >= maxDiffSize)
+      sizesMismatchError = `Expected an image ${expected.width}px by ${expected.height}px, received ${actual.width}px by ${actual.height}px. `;
     actual = resizeImage(actual, size);
     expected = resizeImage(expected, size);
   }
```

Comment:
Allowing for a tolerance level of < (maxDiffSize + 1) to account for floating point error in both CSS and JS.